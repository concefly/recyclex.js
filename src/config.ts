export type IOptions<_K, V> = {
  isEqual: (a?: V, b?: V) => boolean;
  getVersion: (v: V) => string | number;
  useVersionCheck: boolean;
};

export const DefaultOptions: IOptions<string, any> = {
  isEqual: (a, b) => a === b,
  getVersion: () => 0,
  useVersionCheck: false,
};
