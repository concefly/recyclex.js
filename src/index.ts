import {
  map,
  Observable,
  scan,
  Subject,
  BehaviorSubject,
  tap,
  bufferWhen,
  OperatorFunction,
  ObservableInputTuple,
  combineLatest,
  filter,
  takeUntil,
  MonoTypeOperatorFunction,
  Subscription,
  startWith,
} from 'rxjs';

export type IOptions<_K, V> = {
  isEqual: (a?: V, b?: V) => boolean;
  getVersion: (v: V) => string | number;
  useVersionCheck: boolean;
};

export type IOptionsMap<P> = { [K in keyof P]?: Partial<IOptions<K, P[K]>> };

export const DefaultOptions: IOptions<string, any> = {
  isEqual: (a, b) => a === b,
  getVersion: () => 0,
  useVersionCheck: false,
};

export type Blueprint<P extends Record<string, any> = any, R = void> = {
  factory: IComponentFactory<P, R>;
  props: P;
  key: string;
  onInstance?: (ins: IComponentInstance<P, R> | null) => void;
};

export type ITokenDefinition<T> = string & { _type: T };

export type IController = {
  onInit?(): any;
  onBeforeUpdate?(): any;
  onUpdate?(): any;
  onAfterUpdate?(): any;
  onDestroy?(): any;

  onMutateBlueprints?(list: Blueprint[]): Blueprint[];
};

export type IOnBeforeUpdateCB<P extends Record<string, any>> = (props: P, changes: Map<string, any>) => void;
export type IOnUpdateCB<P extends Record<string, any>> = (props: P, changes: Map<string, any>) => Blueprint[] | void;
export type IOnAfterUpdateCB<P extends Record<string, any>> = (props: P, changes: Map<string, any>) => void;

export type IProvideCB = <T>(key: ITokenDefinition<T>, value: T) => T;
export type IInjectCB = <T>(key: ITokenDefinition<T>) => T;

export type IPropSubjects<P extends Record<string, any>> = {
  [K in keyof P as K extends string ? `${K}$` : never]-?: BehaviorSubject<P[K]>;
};

export type IComponentContext<P extends Record<string, any>, R = void> = {
  key: string;

  P: IPropSubjects<P>;
  afterInput$: Subject<Set<keyof P>>;
  afterUpdate$: Subject<void>;
  dispose$: Subject<void>;

  select: <T extends readonly unknown[]>(sources: readonly [...ObservableInputTuple<T>]) => Observable<T>;
  bufferInput: <T>() => OperatorFunction<T, T>;
  takeUntilDispose: <T>() => MonoTypeOperatorFunction<T>;

  addSub: (...subs: Subscription[]) => void;

  provide: IProvideCB;
  inject: IInjectCB;

  setRef: (ref: R) => void;
};

export type ISetupCallback<P extends Record<string, any>, R = void> = (ctx: IComponentContext<P, R>) => Observable<Blueprint[]> | void;

export interface IComponentDefinition<P extends Record<string, any>, R = void> {
  defaultProps: Required<P>;
  setup: ISetupCallback<Required<P>, R>;
  options?: IOptionsMap<Required<P>>;
}

export interface IComponentFactory<P extends Record<string, any>, R = void> {
  def: IComponentDefinition<P, R>;
  create: (
    key: string,
    props?: P,
    parent?: IComponentInstance<any, any>,
    beforeSetup?: (ctx: IComponentContext<P, R>) => void
  ) => IComponentInstance<P, R>;
}

export interface IComponentInstance<P extends Record<string, any>, R = void> {
  key: string;
  parent?: IComponentInstance<any>;

  ref: R;

  contextStore: Map<string, any>;

  provide: IProvideCB;
  inject: IInjectCB;

  update(newProps: P): void;
  dispose(): void;
}

export type IInstanceType<T> = T extends IComponentFactory<infer P, infer R> ? IComponentInstance<P, R> : never;

export class ComponentError<P extends Record<string, any>, R = void> extends Error {
  constructor(
    readonly instance: IComponentInstance<P, R>,
    readonly msg: string,
    readonly innerError?: any
  ) {
    super(msg);
  }
}

export const DefaultError$ = new Subject<ComponentError<any, any>>();
DefaultError$.subscribe(err => {
  console.error(`Error in component ${err.instance.key}: ${err.msg}`);
  console.error(err.innerError);
});

export function blueprint<P extends Record<string, any>, R = void>(
  factory: IComponentFactory<P, R>,
  props: P,
  key: string,
  onInstance?: (ins: IComponentInstance<P, R> | null) => void
): Blueprint {
  // @ts-expect-error
  return { factory, props, key, onInstance };
}

export function defineToken<T>(key: string): ITokenDefinition<T> {
  return key as any;
}

export function defineComponent<P extends Record<string, any>, R = void>(def: IComponentDefinition<P, R>): IComponentFactory<P, R> {
  type _IChild = Blueprint & { ins?: IComponentInstance<any> };

  const create = (
    key: string,
    initProps: P = def.defaultProps,
    parent?: IComponentInstance<any>,
    beforeSetup?: (ctx: IComponentContext<P, R>) => void
  ) => {
    const contextStore = new Map<string, any>();
    const instance: IComponentInstance<P, R> = { key, ref: null as any, contextStore, parent, provide, inject, update, dispose };

    const propSubjects: IPropSubjects<P> = {} as any;
    const inputProps$ = new Subject<P>();

    const afterInput$ = new Subject<Set<keyof P>>();
    const afterUpdate$ = new Subject<void>();
    const dispose$ = new Subject<void>();

    let updating = false;

    const properties = Object.keys(def.defaultProps) as (keyof P)[];

    for (const k of properties) {
      const v = initProps[k] ?? def.defaultProps[k];

      // @ts-expect-error
      propSubjects[`${k}$`] = new BehaviorSubject<any>(v);
    }

    const bufferInput = <T>() => {
      const fn: OperatorFunction<T, T> = src$ =>
        src$.pipe(
          bufferWhen(() => afterInput$),
          filter(args => args.length > 0),
          map(args => args[args.length - 1])
        );

      return fn;
    };

    const select = <T extends readonly unknown[]>(list: readonly [...ObservableInputTuple<T>]) => {
      const vSet = new Set(Object.values(propSubjects));
      const isAllProp = list.every(v => vSet.has(v as any));

      if (!isAllProp) throw new Error('select only support properties stream');

      return combineLatest(list).pipe(bufferInput());
    };

    const takeUntilDispose = <T>() => takeUntil<T>(dispose$);

    const addSub = (...subs: Subscription[]) => {
      for (const sub of subs) {
        dispose$.subscribe(() => sub.unsubscribe());
      }
    };

    const ctx: IComponentContext<P, R> = {
      key,
      P: propSubjects,
      afterUpdate$,
      afterInput$,
      dispose$,
      provide,
      inject,
      bufferInput,
      select,
      takeUntilDispose,
      addSub,
      setRef,
    };

    // setup
    beforeSetup?.(ctx);
    const blueprints$ = def.setup(ctx) ?? new Subject<Blueprint[]>();

    let disposed = false;
    let children: _IChild[] = [];

    const inputSub = inputProps$
      .pipe(
        startWith(initProps), // 初始化输入，给 infos 填充初始值
        scan<P, Record<string, { value: any; version: any; changed: boolean }>>((infos, cur) => {
          for (const key of properties as string[]) {
            const newVal = cur[key] ?? def.defaultProps[key];

            const getVersion = def.options?.[key]?.getVersion ?? DefaultOptions.getVersion;
            const isEqual = def.options?.[key]?.isEqual ?? DefaultOptions.isEqual;
            const useVersionCheck = def.options?.[key]?.useVersionCheck ?? DefaultOptions.useVersionCheck;

            if (!infos[key]) {
              infos[key] = { value: newVal, version: useVersionCheck ? getVersion(newVal) : undefined, changed: false };
              continue;
            }

            let toCompareA: any;
            let toCompareB: any;
            let newVersion: any = undefined;

            if (useVersionCheck) {
              toCompareA = infos[key].version;
              newVersion = getVersion(newVal);
              toCompareB = newVersion;
            } else {
              toCompareA = infos[key].value;
              toCompareB = newVal;
            }

            infos[key].value = newVal;
            infos[key].version = newVersion;
            infos[key].changed = !isEqual(toCompareA, toCompareB);
          }

          return infos;
        }, {}),
        map(infos => {
          const entries = Object.entries(infos).filter(([_, info]) => info.changed);
          if (entries.length === 0) return;

          for (const [k, v] of entries) {
            propSubjects[`${k}$`].next(v.value);
          }

          const changes = new Set<string>(entries.map(([key]) => key));
          return changes;
        }),
        tap(changes => {
          if (changes) afterInput$.next(changes);
        })
      )
      .subscribe({ error: err => DefaultError$.next(new ComponentError(instance, err.message, err)) });

    const updateSub = blueprints$
      .pipe(
        scan(
          (lastInfo, nextChildren) => {
            if (disposed) throw new Error('Already disposed');
            if (updating) throw new Error('Infinity loop detected');

            updating = true;

            const { children: lastChildren, oldChildMap, newChildMap, commonKeys, _toDisposeKeys, _toCreateKeys, _toUpdateKeys } = lastInfo;

            oldChildMap.clear();
            newChildMap.clear();
            commonKeys.clear();

            _toDisposeKeys.clear();
            _toCreateKeys.clear();
            _toUpdateKeys.clear();

            for (const c of lastChildren) oldChildMap.set(c.key, c);
            for (const c of nextChildren) newChildMap.set(c.key, c);

            for (const c of lastChildren) commonKeys.add(c.key);
            for (const c of nextChildren) commonKeys.add(c.key);

            for (let i = 0; i < lastChildren.length; i++) {
              const _n = lastChildren[i];
              if (newChildMap.has(_n.key)) _toUpdateKeys.add(_n.key);
              else _toDisposeKeys.add(_n.key);
            }

            for (let i = 0; i < nextChildren.length; i++) {
              const _n = nextChildren[i];
              if (oldChildMap.has(_n.key)) _toUpdateKeys.add(_n.key);
              else _toCreateKeys.add(_n.key);
            }

            // 卸载
            for (const _k of _toDisposeKeys) {
              const child = oldChildMap.get(_k)!;
              _disposeChild(child);
            }

            // 按顺序创建 or 更新
            for (const child of nextChildren) {
              // create
              if (_toCreateKeys.has(child.key)) {
                _createChild(child);
              }

              // update
              else if (_toUpdateKeys.has(child.key)) {
                const prev = oldChildMap.get(child.key)!;
                const next = newChildMap.get(child.key)!;

                // 类型相同，递归更新
                if (prev.factory === next.factory) {
                  if (!prev.ins) throw new Error('prev.ins not exists');
                  if (!next.ins) next.ins = prev.ins; // 传递实例

                  // set props
                  next.ins.update(next.props);
                }

                // 类型不同，卸载 prev，创建 next
                else {
                  _disposeChild(prev);
                  _createChild(next);
                }
              }
            }

            children = nextChildren;
            lastInfo.children = nextChildren;

            afterUpdate$.next();

            updating = false;
            return lastInfo;
          },
          {
            children: [] as _IChild[],
            oldChildMap: new Map<string, _IChild>(),
            newChildMap: new Map<string, _IChild>(),
            commonKeys: new Set<string>(),
            _toDisposeKeys: new Set<string>(),
            _toCreateKeys: new Set<string>(),
            _toUpdateKeys: new Set<string>(),
          }
        )
      )
      .subscribe({ error: err => DefaultError$.next(new ComponentError(instance, err.message, err)) });

    // blueprint 订阅完成后，立刻触发一次 afterInput，因为 P 是 BehaviorSubject, 会在订阅时立刻发送一次
    afterInput$.next(new Set(properties));

    function _disposeChild(child: _IChild) {
      if (child.ins) {
        child.ins.dispose();
        child.ins = undefined;
        child.onInstance?.(null);
      }
    }

    function _createChild(child: _IChild) {
      child.ins = child.factory.create(child.key, child.props, instance);
      child.onInstance?.(child.ins);
    }

    function provide<T>(key: ITokenDefinition<T>, value: T) {
      if (contextStore.has(key)) throw new Error('key already exists');
      contextStore.set(key, value);
      return value;
    }

    function inject(key: ITokenDefinition<any>) {
      if (contextStore.has(key)) return contextStore.get(key);

      let cur = parent;

      while (cur) {
        if (cur.contextStore.has(key)) {
          return cur.contextStore.get(key);
        }

        cur = cur.parent;
      }

      throw new Error('context not found: ' + key);
    }

    function update(newProps: P) {
      if (disposed) throw new Error('Already disposed');
      inputProps$.next(newProps);
    }

    function setRef(ref: R) {
      instance.ref = ref;
    }

    function dispose() {
      if (disposed) throw new Error('Already disposed');

      for (let i = children.length - 1; i >= 0; i--) {
        children[i].ins?.dispose();
      }

      updateSub.unsubscribe();
      inputSub.unsubscribe();

      // complete
      inputProps$.complete();
      afterUpdate$.complete();
      afterInput$.complete();

      for (const sub of Object.values(propSubjects)) {
        sub.complete();
      }

      dispose$.next();
      dispose$.complete();

      children.length = 0;
      disposed = true;
    }

    return instance;
  };

  return { def, create };
}
