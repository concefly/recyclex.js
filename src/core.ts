import { OperatorFunction, Observable, scan, map, Subject, shareReplay, filter, BehaviorSubject, pipe } from 'rxjs';
import { DefaultOptions, IOptions } from './config';

type ExtractFirst<T> = T extends [infer U, any] ? U : never;
type ExtractSecond<T> = T extends [any, infer U] ? U : never;

export function incremental<T, LS extends [key: string, value: any][]>(
  split: (src: T) => LS,
  options?: Record<string, Partial<IOptions<string, any>>> | ((key: string) => Partial<IOptions<string, any>>)
): OperatorFunction<T, { [K in keyof LS]: Observable<ExtractSecond<LS[K]>> }> {
  return src$ => {
    let lastStreams: Map<string, { subject: Subject<any>; ob: Observable<any> }> = new Map();

    return new Observable(sub => {
      const inputSub = src$
        .pipe(
          map(list => {
            const oldGroups = lastStreams;
            const newGroups = new Map<string, T>();

            const listEntries = split(list);

            for (const [k, v] of listEntries) {
              if (newGroups.has(k)) throw new Error('keyBy must be unique, found duplicate key: ' + k);
              newGroups.set(k, v);
            }

            const toRemoveKeys = new Set<string>();
            const toCreateKeys = new Set<string>();
            const toUpdateKeys = new Set<string>();

            for (const k of oldGroups.keys()) {
              if (!newGroups.has(k)) toRemoveKeys.add(k);
              else toUpdateKeys.add(k);
            }

            for (const k of newGroups.keys()) {
              if (!oldGroups.has(k)) toCreateKeys.add(k);
            }

            const streams = new Map<string, { subject: Subject<any>; ob: Observable<any> }>();
            const orderedObs: Observable<any>[] = [];
            let streamsChanged = false;

            // remove old
            for (const k of toRemoveKeys) {
              const { subject } = oldGroups.get(k)!;
              subject.complete();
              streamsChanged = true;
            }

            // 按顺序创建 or 更新
            for (const [key, value] of listEntries) {
              // create
              if (toCreateKeys.has(key)) {
                const subject = new BehaviorSubject<T>(value);

                const ob = subject.pipe(
                  scan(
                    (lastInfos, newVal) => {
                      const getVersion =
                        (options ? (typeof options === 'function' ? options(key).getVersion : options[key]?.getVersion) : undefined) ??
                        DefaultOptions.getVersion;

                      const isEqual =
                        (options ? (typeof options === 'function' ? options(key).isEqual : options[key]?.isEqual) : undefined) ??
                        DefaultOptions.isEqual;

                      const useVersionCheck =
                        (options
                          ? typeof options === 'function'
                            ? options(key).useVersionCheck
                            : options[key]?.useVersionCheck
                          : undefined) ?? DefaultOptions.useVersionCheck;

                      let toCompareA: any;
                      let toCompareB: any;
                      let newVersion: any = undefined;

                      if (useVersionCheck) {
                        toCompareA = lastInfos.version;
                        newVersion = getVersion(newVal);
                        toCompareB = newVersion;
                      } else {
                        toCompareA = lastInfos.value;
                        toCompareB = newVal;
                      }

                      return { value: newVal, version: newVersion, changed: !isEqual(toCompareA, toCompareB) };
                    },
                    { value: null as any, version: null as any, changed: false }
                  ),
                  filter(({ changed }) => changed),
                  map(t => t.value),
                  shareReplay(1)
                );
                ob.subscribe();

                streams.set(key, { subject, ob });
                orderedObs.push(ob);

                streamsChanged = true;
              }

              // update
              else if (toUpdateKeys.has(key)) {
                const { subject, ob } = oldGroups.get(key)!;

                subject.next(value);
                streams.set(key, { subject, ob });
                orderedObs.push(ob);
              }
            }

            lastStreams = streams;

            return { orderedObs, streams, streamsChanged };
          }),
          filter(({ streamsChanged }) => streamsChanged),
          map(({ orderedObs }) => orderedObs)
        )
        .subscribe({
          next: obs => sub.next(obs as any),
          error: err => sub.error(err),
          complete: () => sub.complete(),
        });

      return () => {
        inputSub.unsubscribe();
        lastStreams.forEach(({ subject }) => subject.complete());
      };
    });
  };
}

export function incrementalList<T>(
  keyBy: (item: T) => string,
  options?: Partial<IOptions<string, T>>
): OperatorFunction<T[], Observable<T>[]> {
  return pipe(
    incremental(
      list => {
        return list.map(item => [keyBy(item), item] as const);
      },
      () => options || DefaultOptions
    )
  );
}
