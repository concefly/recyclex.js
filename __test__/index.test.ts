import { it, expect } from 'vitest';
import { blueprint, DefaultError$, defineComponent, defineContext, IInstanceType } from '../src';
import { BehaviorSubject, combineLatest, map, take } from 'rxjs';

it('Single Component Lifecycle', () => {
  const timelines: string[] = [];

  const A = defineComponent({
    defaultProps: { n: 1 },
    setup: ctx => {
      timelines.push(`a init`);

      ctx.afterUpdate$.subscribe(() => {
        timelines.push('a afterUpdate');
      });

      ctx.dispose$.subscribe(() => {
        timelines.push('a dispose');
      });

      return combineLatest([ctx.P.n$]).pipe(
        map(([n]) => {
          timelines.push(`a update: n=${n}`);
          return [];
        })
      );
    },
  });

  const a = A.create('root');

  a.update({ n: 2 });
  a.update({ n: 2 });

  a.update({ n: 3 });
  a.update({ n: 3 });

  a.dispose();

  expect(timelines).toMatchInlineSnapshot(`
    [
      "a init",
      "a update: n=1",
      "a afterUpdate",
      "a update: n=2",
      "a afterUpdate",
      "a update: n=3",
      "a afterUpdate",
      "a dispose",
    ]
  `);
});

it('Single Component Lifecycle with initProps', () => {
  const timelines: string[] = [];

  const A = defineComponent({
    defaultProps: { n: 1 },
    setup: ctx => {
      timelines.push(`a init`);

      ctx.dispose$.subscribe(() => {
        timelines.push('a dispose');
      });

      return combineLatest([ctx.P.n$]).pipe(
        map(([n]) => {
          timelines.push(`a update: n=${n}`);
          return [];
        })
      );
    },
  });

  const a = A.create('root', { n: 2 });
  a.dispose();

  expect(timelines).toMatchInlineSnapshot(`
    [
      "a init",
      "a update: n=2",
      "a dispose",
    ]
  `);
});

it('Cannot dispose twice', () => {
  const A = defineComponent({
    defaultProps: { n: 1 },
    setup: () => {},
  });

  const a = A.create('root');

  a.dispose();

  expect(() => a.dispose()).toThrow('Already disposed');
});

it('Cannot update after dispose', () => {
  const A = defineComponent({
    defaultProps: { n: 1 },
    setup: () => {},
  });

  const a = A.create('root');

  a.dispose();

  expect(() => a.update({ n: 2 })).toThrow('Already disposed');
});

it('Hierarchy Components Lifecycle', () => {
  const timelines: string[] = [];

  const B = defineComponent({
    defaultProps: { b: 1 },
    setup: ctx => {
      timelines.push(`b<${ctx.key}> init`);

      ctx.dispose$.subscribe(() => {
        timelines.push(`b<${ctx.key}> dispose`);
      });

      return combineLatest([ctx.P.b$]).pipe(
        map(([n]) => {
          timelines.push(`b<${ctx.key}> update: b=${n}`);
          return [];
        })
      );
    },
  });

  const A = defineComponent({
    defaultProps: { a: 1 },
    setup: ctx => {
      timelines.push(`a<${ctx.key}> init`);

      ctx.dispose$.subscribe(() => {
        timelines.push(`a<${ctx.key}> dispose`);
      });

      return combineLatest([ctx.P.a$]).pipe(
        map(([a]) => {
          timelines.push(`a<${ctx.key}> update: a=${a}`);

          if (a === 1) return [blueprint(B, { b: 1 }, '1')];
          if (a === 2) return [blueprint(B, { b: 1.1 }, '1'), blueprint(B, { b: 2 }, '2')];
          if (a === 3) return [blueprint(B, { b: 2.1 }, '2')];

          return [];
        })
      );
    },
  });

  const root = A.create('root');

  root.update({ a: 2 });
  root.update({ a: 3 });

  root.dispose();

  expect(timelines).toMatchInlineSnapshot(`
    [
      "a<root> init",
      "a<root> update: a=1",
      "b<1> init",
      "b<1> update: b=1",
      "a<root> update: a=2",
      "b<1> update: b=1.1",
      "b<2> init",
      "b<2> update: b=2",
      "a<root> update: a=3",
      "b<1> dispose",
      "b<2> update: b=2.1",
      "b<2> dispose",
      "a<root> dispose",
    ]
  `);
});

it('Hierarchy Components Context', () => {
  const timelines: string[] = [];

  const userContext = defineContext<BehaviorSubject<string>>('user');

  const B = defineComponent({
    defaultProps: { b: 1 },
    setup: ctx => {
      timelines.push(`b<${ctx.key}> init`);

      const user = ctx.getContext(userContext);
      timelines.push(`b<${ctx.key}> user: ${user.value}`);

      ctx.dispose$.subscribe(() => {
        timelines.push(`b<${ctx.key}> dispose`);
      });

      return combineLatest([ctx.P.b$]).pipe(
        map(() => {
          timelines.push(`b<${ctx.key}> update user: ${user.value}`);
          return [];
        })
      );
    },
  });

  const A = defineComponent({
    defaultProps: { a: 1 },
    setup: ctx => {
      timelines.push(`a<${ctx.key}> init`);

      const user = ctx.createContext(userContext, new BehaviorSubject('xxx'));
      timelines.push(`a<${ctx.key}> user: ${user.value}`);

      user.next('user_1');

      ctx.dispose$.subscribe(() => {
        timelines.push(`a<${ctx.key}> dispose`);
      });

      return combineLatest([ctx.P.a$]).pipe(
        map(([a]) => {
          if (a === 2) {
            user.next('user_2');
          }

          return [blueprint(B, { b: a }, '1')];
        })
      );
    },
  });

  const root = A.create('root');

  root.update({ a: 2 });
  root.dispose();

  expect(timelines).toMatchInlineSnapshot(`
    [
      "a<root> init",
      "a<root> user: xxx",
      "b<1> init",
      "b<1> user: user_1",
      "b<1> update user: user_1",
      "b<1> update user: user_2",
      "b<1> dispose",
      "a<root> dispose",
    ]
  `);
});

it('Hierarchy Components Context Subscribe', async () => {
  const timelines: string[] = [];

  const userContext = defineContext<BehaviorSubject<string>>('user');

  const B = defineComponent({
    defaultProps: {},
    setup: ctx => {
      const userContextVal$ = ctx.getContext(userContext);

      return userContextVal$.pipe(
        map(() => {
          timelines.push(`b<${ctx.key}> update user: ${userContextVal$.value}`);
          return [];
        })
      );
    },
  });

  const A = defineComponent({
    defaultProps: { a: 1 },
    setup: ctx => {
      const userContextVal = ctx.createContext(userContext, new BehaviorSubject('xxx'));

      setTimeout(() => {
        userContextVal.next('user_1');
        userContextVal.next('user_2');
      }, 0);

      return combineLatest([ctx.P.a$]).pipe(
        map(() => {
          return [blueprint(B, {}, '1')];
        })
      );
    },
  });

  const root = A.create('root');

  await new Promise<void>(resolve => setTimeout(resolve, 100));

  root.dispose();

  expect(timelines).toMatchInlineSnapshot(`
    [
      "b<1> update user: xxx",
      "b<1> update user: user_1",
      "b<1> update user: user_2",
    ]
  `);
});

it('Missing Key', () => {
  const timelines: string[] = [];

  const A = defineComponent<{ n: number; k?: string }>({
    defaultProps: { n: 1, k: '_default_' },
    setup: ctx => {
      return combineLatest([ctx.P.n$, ctx.P.k$]).pipe(
        map(([n, k]) => {
          timelines.push(`a update: n=${n}, k=${k}`);
          return [];
        })
      );
    },
  });

  const a = A.create('root');

  a.update({ n: 2, k: 'a' });
  a.update({ n: 3 });
  a.update({ n: 4, k: 'b' });

  a.dispose();

  expect(timelines).toMatchInlineSnapshot(`
    [
      "a update: n=1, k=_default_",
      "a update: n=2, k=_default_",
      "a update: n=2, k=a",
      "a update: n=3, k=a",
      "a update: n=3, k=_default_",
      "a update: n=4, k=_default_",
      "a update: n=4, k=b",
    ]
  `);
});

it('bufferInput', () => {
  const timelines: string[] = [];

  const A = defineComponent<{ n: number; k?: string }>({
    defaultProps: { n: 1, k: '_default_' },
    setup: ctx => {
      return combineLatest([ctx.P.n$, ctx.P.k$]).pipe(
        ctx.bufferInput(),
        map(([n, k]) => {
          timelines.push(`a update: n=${n}, k=${k}`);
          return [];
        })
      );
    },
  });

  const a = A.create('root');

  a.update({ n: 2, k: 'a' });
  a.update({ n: 3 });
  a.update({ n: 4, k: 'b' });

  a.dispose();

  expect(timelines).toMatchInlineSnapshot(`
    [
      "a update: n=1, k=_default_",
      "a update: n=2, k=a",
      "a update: n=3, k=_default_",
      "a update: n=4, k=b",
    ]
  `);
});

it('props select', () => {
  const timelines: string[] = [];

  const A = defineComponent<{ n: number; k?: string }>({
    defaultProps: { n: 1, k: '_default_' },
    setup: ctx => {
      return ctx.select([ctx.P.n$, ctx.P.k$]).pipe(
        map(([n, k]) => {
          timelines.push(`a update: n=${n}, k=${k}`);
          return [];
        })
      );
    },
  });

  const a = A.create('root');

  a.update({ n: 2, k: 'a' });
  a.update({ n: 3 });
  a.update({ n: 4, k: 'b' });

  a.dispose();

  expect(timelines).toMatchInlineSnapshot(`
    [
      "a update: n=1, k=_default_",
      "a update: n=2, k=a",
      "a update: n=3, k=_default_",
      "a update: n=4, k=b",
    ]
  `);
});

it('Single Component Ref', () => {
  const A = defineComponent<{ n: number }, { name: string }>({
    defaultProps: { n: 1 },
    setup: ctx => {
      ctx.afterInput$.subscribe(() => {
        ctx.setRef({ name: ctx.P.n$.value + '' });
      });
    },
  });

  const a = A.create('root');
  expect(a.ref.name).toBe('1');

  a.update({ n: 2 });
  expect(a.ref.name).toBe('2');

  a.dispose();
});

it('Hierarchy Components Ref', () => {
  const timelines: string[] = [];

  const B = defineComponent<{ b: number }, { name: string }>({
    defaultProps: { b: 1 },
    setup: ctx => {
      ctx.P.b$.subscribe(b => {
        ctx.setRef({ name: b + '' });
      });
    },
  });

  const A = defineComponent({
    defaultProps: { a: 1 },
    setup: ctx => {
      let bIns: IInstanceType<typeof B> | null = null;

      ctx.afterUpdate$.subscribe(() => {
        timelines.push(`a afterUpdate: a=${ctx.P.a$.value}, bKey=${bIns?.key}, bRef=${bIns?.ref?.name}`);
      });

      return combineLatest([ctx.P.a$]).pipe(
        map(([a]) => {
          if (a <= 0) return [];
          return [blueprint(B, { b: a }, Math.floor(a) + '', ins => (bIns = ins))];
        })
      );
    },
  });

  const root = A.create('root');

  root.update({ a: 1.1 });
  root.update({ a: 2 });
  root.update({ a: 0 });
  root.dispose();

  expect(timelines).toMatchInlineSnapshot(`
    [
      "a afterUpdate: a=1, bKey=1, bRef=1",
      "a afterUpdate: a=1.1, bKey=1, bRef=1.1",
      "a afterUpdate: a=2, bKey=2, bRef=2",
      "a afterUpdate: a=0, bKey=undefined, bRef=undefined",
    ]
  `);
});

it('Infinity Loop', async () => {
  return new Promise<void>(resolve => {
    const A = defineComponent<{ n: number }, { name: string }>({
      defaultProps: { n: 1 },
      setup: ctx => {
        const tick$ = new BehaviorSubject(0);

        ctx.afterUpdate$.subscribe(() => {
          tick$.next(tick$.value + 1);
        });

        return tick$.pipe(
          map(() => {
            return [];
          })
        );
      },
    });

    DefaultError$.pipe(take(1)).subscribe(err => {
      expect(err.message).toContain('Infinity loop detected');
      resolve();
    });

    A.create('root');
  });
});
