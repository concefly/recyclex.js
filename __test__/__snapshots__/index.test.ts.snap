// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`equals check 1`] = `
[
  "<A> receive: name=TOM,",
  "<A> receive: name=JANE,",
]
`;

exports[`life cycle > with initial values 1`] = `
[
  "<A> onInit name=Jam, color=red,",
  "<A> onBeforeUpdate name=Jam, color=red,",
  "<A> onUpdate name=Jam, color=red,",
  "  <B> onInit text=Jam-red,",
  "  <B> onBeforeUpdate text=Jam-red,",
  "  <B> onUpdate text=Jam-red,",
  "  <B> onAfterUpdate text=Jam-red,",
  "<A> onAfterUpdate name=Jam, color=red,",
]
`;

exports[`life cycle 1`] = `
[
  "<A> onInit name=Jam, color=red,",
  "<A> onBeforeUpdate name=Jam, color=red,",
  "<A> onUpdate name=Jam, color=red,",
  "  <B> onInit text=Jam-red,",
  "  <B> onBeforeUpdate text=Jam-red,",
  "  <B> onUpdate text=Jam-red,",
  "  <B> onAfterUpdate text=Jam-red,",
  "<A> onAfterUpdate name=Jam, color=red,",
  "<A> onBeforeUpdate name=Tom, color=red,",
  "<A> onUpdate name=Tom, color=red,",
  "  <B> onBeforeUpdate text=Tom-red,",
  "  <B> onUpdate text=Tom-red,",
  "  <B> onAfterUpdate text=Tom-red,",
  "<A> onAfterUpdate name=Tom, color=red,",
  "<A> onBeforeUpdate name=Jane, color=red,",
  "<A> onUpdate name=Jane, color=red,",
  "  <B> onBeforeUpdate text=Jane-red,",
  "  <B> onUpdate text=Jane-red,",
  "  <B> onAfterUpdate text=Jane-red,",
  "<A> onAfterUpdate name=Jane, color=red,",
  "  <B> onDestroy text=Jane-red,",
  "<A> onDestroy name=Jane, color=red,",
]
`;

exports[`life cycle: with initial values 1`] = `
[
  "<A> onInit name=Jam, color=red,",
  "<A> onBeforeUpdate name=Jam, color=red,",
  "<A> onUpdate name=Jam, color=red,",
  "  <B> onInit text=Jam-red,",
  "  <B> onBeforeUpdate text=Jam-red,",
  "  <B> onUpdate text=Jam-red,",
  "  <B> onAfterUpdate text=Jam-red,",
  "<A> onAfterUpdate name=Jam, color=red,",
]
`;
